server/channels/jobs/batch_migration_worker_test.go

```diff
@@ -288,6 +288,7 @@ func TestBatchMigrationWorker(t *testing.T) {
 			require.Equal(t, 1, batchNumber, "only batch 1 should have run")
 
 			if batchNumber >= 1 {
+				go worker.Stop()
```

Comment:
@lieut-data I made some changes to the tests because I noticed that we weren't actually running the correct number of batches in a couple cases, but it also revealed that for some reason the worker won't stop running unless we explicitly run `worker.Stop()`.

In these tests, this shouldn't be required since the job server should stop them once we tell the batch migration worker that it's done, but a later cleanup step fails without these. Is this just a limitation of the tests (ie. the job server is supposed to do this) or do we have an issue with the batch worker?

Reply:
> it also revealed that for some reason the worker won't stop running unless we explicitly run worker.Stop().

I think this is expected, given the worker's Run loop either waiting for new jobs or the stop signal:
https://github.com/mattermost/mattermost/blob/f72aed2263a986becf69fe9a98cccc1e40c7d0cd/server/channels/jobs/batch_migration_worker.go#L78-L87

> In these tests, this shouldn't be required since the job server should stop them once we tell the batch migration worker that it's done

I'm not sure this is true: isn't a worker supposed to remain alive indefinitely to process a new job, only shutting down when the server stops?

> but a later cleanup step fails without these

Can you elaborate on what exactly is failing here?

Reply:
Okay so I am misunderstanding the concept of the "worker" here, I think I'm misconstruing it with an actual "job". So what should be happening then is that the worker should only be shutting down when sent the `stop` signal, meaning that I think we need to remove the `waitDone` check for every single test here, since not all of them should be shutting down the worker.

Reply:
> So what should be happening then is that the worker should only be shutting down when sent the stop signal, meaning that I think we need to remove the `waitDone` check for every single test here, since not all of them should be shutting down the worker.

I might be misremembering, but I believe `waitDone` is being used to assert the total behaviour of the worker by asserting that it stops vs. accidentally passing a test while the worker actually plods on ahead and tries to do more on the job than would be expected. Unless there's something demonstrably wrong, I think `waitDone` is working as designed.

Reply:
Yes I ended up keeping it as is so that should clear it all up.